### PR TITLE
[#251] Added Deep Learning implementation for solving regression

### PR DESCRIPTION
**Pull request by arekmano**
[#251]

Implementation of a regressor to predict the tenant_ordered_to_pay_landlord outcome. Uses a neural network implementation to achieve this.

### CODE REVIEW

**Comment by arekmano** [comment]
Results using Neural Net
=====================

Standardized: -51.57 (5.92) MSE, model: intermed_model, epochs: 50
Standardized: -27.62 (7.40) MSE, model: adv_model, epochs: 50
Standardized: -22.65 (3.44) MSE, model: vadv_model, epochs: 50
Standardized: -26.38 (3.94) MSE, model: intermed_model, epochs: 100
Standardized: -20.34 (2.31) MSE, model: adv_model, epochs: 100
Standardized: -19.19 (2.00) MSE, model: vadv_model, epochs: 100

**Comment by arekmano** [comment]
Using a basic SVM to classify Win/No Win
===

Size of dataset: 15227
Sample size: 15227
Train size: 12181
Test size: 3046

LINEAR
Test accuracy: 62.17990807616546%
Precision: [ 0.61008403 0.624643 ]
Recall: [ 0.28293063 0.86840613]
F1: [ 0.38658147 0.72662553]

BEST PARAMS
{'C': 5, 'kernel': 'rbf'}

RBF, C=5
Test accuracy: 62.77084701247538%
Precision: [ 0.68350168 0.6141925 ]
Recall: [ 0.30029586 0.88902007]
F1: [ 0.41726619 0.72648336]

RBF, C=7
Test accuracy: 63.591595535128036%
Precision: [ 0.65693431 0.62981787]
Recall: [ 0.33987915 0.86353078]
F1: [ 0.44798407 0.72838599]

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        np.random.seed(seed)
+        X = np.array([precedent['facts_vector'] for precedent in self.dataset])
+        Y = np.array([precedent['outcomes_vector'][10]
+                      for precedent in self.dataset])
+        kfold = KFold(n_splits=10, random_state=seed)
+        results = cross_val_score(self.model, X, Y, cv=kfold)
+        Log.write("Mean: %.2f (%.2f) MSE" %
+                  (results.mean(), results.std()))
+
+    def predict(self, precedent):
```
Comment (Samuel-Campbell) [inline src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py:120]: you should scale the data before predicting

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                      for precedent in self.dataset])
+        regressor = KerasRegressor(
+            build_fn=TenantPaysLandlordRegressor.__adv_model, epochs=100, batch_size=128, verbose=0)
+        scaler = StandardScaler()
+        self.model = TenantPaysLandlordRegressor.__create_architecture(scaler, regressor)
+        self.model.fit(X, Y)
+        self.dataset = None
+        self.save()
+
+    def __create_architecture(scaler, regressor):
```
Comment (Samuel-Campbell) [inline src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: @staticmethod

Comment (Samuel-Campbell) [reply src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: scaler not being used

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            ...
+        }]
+        """
+        if dataset is not None:
+            self.dataset = [precedent for precedent in dataset.values() if precedent[
+                'outcomes_vector'][10] > 1]
+        else:
+            self.load()
+
+    def __adv_model():
```
Comment (Samuel-Campbell) [inline src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: @staticmethod

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py (hunk 1 above)
Comment (arekmano) [reply src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py:120]: Pipeline() uses all the operators found in its pipeline. In this case, it is both the scaler and the regressor.

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py (hunk 3 above)
Comment (arekmano) [reply src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: fixing

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py (hunk 2 above)
Comment (arekmano) [reply src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: good catch

**Review by Samuel-Campbell — APPROVED** [review]
